### PR TITLE
Rebalanced Fire v2

### PR DESCRIPTION
**Pull request by NotFuji**
Because I accidentally deleted the other one.

**Anyway...**
 - Fire now progresses in 4 stages, the smallest represents a candle flame or lighter, other three are the same as the existing three.  To compensate, fire based weapons start stage 2 fires rather than stage 1.  Materials are adjusted appropriately, and some outright won't burn unless put in a fire of stage 2+.

 - _Start a fire quickly_ action requires tinder to make a sustainable fire on most materials.  Tinder can either be small flammable items on the tile you're trying to light, or 25 charges of the "tinder" item, which is automatically consumed from the player's inventory if they possess enough.  The player can still start small fires without tinder if they choose to, but as the fire will likely die out it's probably a bad idea to do so.
_See **item::is_tinder()** for more information on what does or doesn't count as tinder._

 - Move cost of _Start a fire quickly_ action is determined by the highest fuel rating of the flammable items on a tile.  With a "lighter", wood will take about 1 minute to ignite, while gasoline ignites near instantaneously.

 - Fires on terrains and furniture are far less destructive.  Large fires in structures are unlikely to destroy the entire building, however damage will still be substantial.

### CODE REVIEW

**Comment by NotFuji** [comment]
They get an entry in their description saying it will work as tinder.
paper

**Comment by Leland** [comment]
Haven't tested to see if this was pre-existing:

If you shoot a flamethrower at a building wall, it seems to instantly destroy it and dump debris everywhere

screen recording 2018-05-15 at 08 08 pm

**Comment by NotFuji** [comment]
Lowered the chance for smaller fires to destroy terrains, so while it might still happen, there's only a 0.7 - 1% chance per tick of it happening.

File: src/field.cpp [hunk 1]
```diff
@@ -927,47 +927,26 @@ bool map::process_fields_in_submap( submap *const current_submap,
...
-
-                            } else if( ter_furn_has_flag( ter, frn, TFLAG_FLAMMABLE_HARD ) &&
-                                       one_in( 3 ) ) {
-                                // The fire feeds on the ground itself until max density.
-                                time_added += 1_turns * ( 4 - cur.getFieldDensity() );
                                 smoke += 2;
-                                if( cur.getFieldDensity() > 1 &&
-                                    one_in( 200 - cur.getFieldDensity() * 50 ) ) {
+                                if( cur.getFieldDensity() >= FIRE_SIZE_THRESHOLD &&
+                                    one_in( 250 - ( cur.getFieldDensity() ) * 50 ) ) {
```
Comment (BevapDin) [inline src/field.cpp]: Why the `()` around `cur.getFieldDensity()`?

File: src/item.cpp [hunk 2]
```diff
@@ -4903,6 +4907,28 @@ bool item::reload( player &u, item_location loc, long qty )
...
+    float fuel = 0.0f;
+    for( const auto &m : mats ) {
+        const auto &bd = m->burn_data( 1 );
+        if( bd.immune ) {
+            break;
+        }
+        fuel += bd.fuel;
+    }
+
+    float vol = base_volume().value() / 1000.0f;
```
Comment (BevapDin) [inline src/item.cpp]: Please use something like `to_milliliter( base_volume() )` so a reader has a better understanding what range to expect here. And maybe change the name of `vol` to `vol_in_liter`.

File: src/iuse_actor.cpp [hunk 3]
```diff
@@ -973,27 +972,59 @@ bool firestarter_actor::prep_firestarter_use( const player &p, tripoint &pos )
...
         if( has_unactivated_brazier &&
             !query_yn(_("There's a brazier there but you haven't set it up to contain the fire. Continue?")) ) {
             return false;
         }
-        return true;
+
+        inventory &inv = g->u.inv;
+        item &tinder = inv.find_item( inv.position_by_type( "tinder" ) );
+        if( has_tinder_here ) {
+            *density = 2;
```
Comment (BevapDin) [inline src/iuse_actor.cpp]: As you seem to use `density` without checking it for being null, you pretty much require callers to supply a valid pointer as argument. Why is it a pointer and not a reference?

File: src/iuse_actor.cpp [hunk 4]
```diff
@@ -973,27 +972,59 @@ bool firestarter_actor::prep_firestarter_use( const player &p, tripoint &pos )
...
+            }
         }
         if( has_unactivated_brazier &&
             !query_yn(_("There's a brazier there but you haven't set it up to contain the fire. Continue?")) ) {
             return false;
         }
-        return true;
+
+        inventory &inv = g->u.inv;
+        item &tinder = inv.find_item( inv.position_by_type( "tinder" ) );
```
Comment (BevapDin) [inline src/iuse_actor.cpp]: As you're using `tinder` only inside one of the branches below, it should only be created there (as close to its usage as possible).

File: src/iuse_actor.cpp [hunk 5]
```diff
@@ -973,27 +972,59 @@ bool firestarter_actor::prep_firestarter_use( const player &p, tripoint &pos )
...
+            return true;
+        } else if( inv.has_tools( "tinder", 1 ) ) {
+            int used_charges = tinder.type->stack_size / 2;
+            if( tinder.count_by_charges() && tinder.charges >= used_charges ) {
+                tinder.charges -= used_charges;
+                if( tinder.charges <= 0 ) {
+                    inv.remove_item( &tinder );
+                }
+
+                p.add_msg_if_player( m_info, string_format( ( "You use %d charges of tinder to light the fire." ),
```
Comment (BevapDin) [inline src/iuse_actor.cpp]: Why are there `()` around the string literal. And it should be translated. Maybe only the `_` is missing? Same below.

File: src/fire.h [hunk 6]
```diff
@@ -2,6 +2,9 @@
 #ifndef FIRE_H
 #define FIRE_H
 
+const int MAX_FIRE_SIZE = 4;
```
Comment (BevapDin) [inline src/fire.h]: Have you tested this carefully? We currently only support field density up to 3 (`MAX_FIELD_DENSITY` is 3) and the fields data (inside of `field_t`) has only data for that many density levels.

A field density of more than 3 will lead to errors or just undefined behaviour.

Comment (NotFuji) [reply src/fire.h]: Resolved: see commit 1d72fd0

**Comment by ZhilkinSerg** [comment]
Can you resolve conflicts, please?

**Comment by NotFuji** [comment]
I'll probably end up redoing this in a way that's more compatible with the current field's system (and more reliable overall).  And split the lighter rebalance into a separate PR.  It works as-is, but I'm not quite sure if it's fun or balanced at all, nor do I have the time currently to find out.

**Comment by ZhilkinSerg** [comment]
> I'll probably end up redoing this in a way that's more compatible with the current field's system (and more reliable overall). And split the lighter rebalance into a separate PR. It works as-is, but I'm not quite sure if it's fun or balanced at all, nor do I have the time currently to find out.

Okay. I will mark it as `Stalled` and close to clean up PR queue, but feel free to reopen it.

**Pull request closed without merging**